src/erasure-code/shec/ErasureCodeShec.cc

```diff
@@ -197,7 +197,10 @@ int ErasureCodeShec::_decode(const set<int> &want_to_read,
   }
   unsigned int k = get_data_chunk_count();
   unsigned int m = get_chunk_count() - k;
-  unsigned blocksize = (*chunks.begin()).second.length();
+  unsigned blocksize = 0;
```

Comment:
Sure, wilco. But it would take sometime - armv7l is extremely slow.

Reply:
Here's what I got with ceph_assert(!chunks.empty()) . I'll do gdb backtrace for clarity shortly.

```
[root@blank build]# ctest -R unittest_erasure_code_shec_arguments -V
UpdateCTestConfiguration  from :/mnt/git/debug/ceph-15.2.4/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/mnt/git/debug/ceph-15.2.4/build/DartConfiguration.tcl
Test project /mnt/git/debug/ceph-15.2.4/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 124
    Start 124: unittest_erasure_code_shec_arguments

124: Test command: /mnt/git/debug/ceph-15.2.4/build/bin/unittest_erasure_code_shec_arguments
124: Environment variables: 
124:  CEPH_ROOT=/mnt/git/debug/ceph-15.2.4
124:  CEPH_BIN=/mnt/git/debug/ceph-15.2.4/build/bin
124:  CEPH_LIB=/mnt/git/debug/ceph-15.2.4/build/lib
124:  CEPH_BUILD_DIR=/mnt/git/debug/ceph-15.2.4/build
124:  LD_LIBRARY_PATH=/mnt/git/debug/ceph-15.2.4/build/lib
124:  PATH=/mnt/git/debug/ceph-15.2.4/build/bin:/mnt/git/debug/ceph-15.2.4/src:/bin:/usr/bin:/sbin:/usr/sbin
124:  PYTHONPATH=/mnt/git/debug/ceph-15.2.4/build/lib/cython_modules/lib.3:/mnt/git/debug/ceph-15.2.4/src/pybind
124:  CEPH_BUILD_VIRTUALENV=/mnt/git/debug/ceph-15.2.4/build
124: Test timeout computed to be: 3600
124: did not load config file, using default settings.
124: 2020-11-25T07:14:04.539+0000 b6f1f010 -1 Errors while parsing config file!
124: 2020-11-25T07:14:04.539+0000 b6f1f010 -1 parse_file: filesystem error: cannot get file size [ceph.conf]
124: 2020-11-25T07:14:04.539+0000 b6f1f010 -1 Errors while parsing config file!
124: 2020-11-25T07:14:04.539+0000 b6f1f010 -1 parse_file: filesystem error: cannot get file size [ceph.conf]
124: [==========] Running 1 test from 1 test suite.
124: [----------] Global test environment set-up.
124: [----------] 1 test from ParameterTest
124: [ RUN      ] ParameterTest.combination_all
124: shec_make_decoding_matrix(): can't find recover matrix.
124: /mnt/git/debug/ceph-15.2.4/src/erasure-code/shec/ErasureCodeShec.cc: In function 'virtual int ErasureCodeShec::_decode(const std::set<int>&, const std::map<int, ceph::buffer::v15_2_0::list>&, std::map<int, ceph::buffer::v15_2_0::list>*)' thread b6f1f010 time 2020-11-25T07:14:04.613041+0000
124: /mnt/git/debug/ceph-15.2.4/src/erasure-code/shec/ErasureCodeShec.cc: 200: FAILED ceph_assert(!chunks.empty())
124:  ceph version 15.2.4 (7447c15c6ff58d7fce91843b705a268a1917325c) octopus (stable)
124:  1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x190) [0xb1eb8df4]
124:  2: (()+0x1551090) [0xb1eb9090]
124:  3: (ErasureCodeShec::_decode(std::set<int, std::less<int>, std::allocator<int> > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > >*)+0x2a4) [0xb09277f0]
124:  4: (ceph::ErasureCode::decode(std::set<int, std::less<int>, std::allocator<int> > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > >*, int)+0x40) [0xb090ec44]
124:  5: (ParameterTest_combination_all_Test::TestBody()+0x14f4) [0x560960]
124:  6: (void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x74) [0x5bbeac]
124:  7: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x68) [0x5b0f44]
124:  8: (testing::Test::Run()+0x110) [0x5848b0]
124:  9: (testing::TestInfo::Run()+0x140) [0x585368]
124:  10: (testing::TestSuite::Run()+0x14c) [0x585c60]
124:  11: (testing::internal::UnitTestImpl::RunAllTests()+0x480) [0x5929c0]
124:  12: (bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x74) [0x5bd978]
124:  13: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x68) [0x5b2734]
124:  14: (testing::UnitTest::Run()+0xe4) [0x590f84]
124:  15: (RUN_ALL_TESTS()+0x18) [0x5652ec]
124:  16: (main()+0xb8) [0x563248]
124:  17: (__libc_start_main()+0x114) [0xb01d0848]
124: *** Caught signal (Aborted) **
124:  in thread b6f1f010 thread_name:unittest_erasur
124: 2020-11-25T07:14:04.634+0000 b6f1f010 -1 /mnt/git/debug/ceph-15.2.4/src/erasure-code/shec/ErasureCodeShec.cc: In function 'virtual int ErasureCodeShec::_decode(const std::set<int>&, const std::map<int, ceph::buffer::v15_2_0::list>&, std::map<int, ceph::buffer::v15_2_0::list>*)' thread b6f1f010 time 2020-11-25T07:14:04.613041+0000
124: /mnt/git/debug/ceph-15.2.4/src/erasure-code/shec/ErasureCodeShec.cc: 200: FAILED ceph_assert(!chunks.empty())
124: 
124:  ceph version 15.2.4 (7447c15c6ff58d7fce91843b705a268a1917325c) octopus (stable)
124:  1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x190) [0xb1eb8df4]
124:  2: (()+0x1551090) [0xb1eb9090]
124:  3: (ErasureCodeShec::_decode(std::set<int, std::less<int>, std::allocator<int> > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > >*)+0x2a4) [0xb09277f0]
124:  4: (ceph::ErasureCode::decode(std::set<int, std::less<int>, std::allocator<int> > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > >*, int)+0x40) [0xb090ec44]
124:  5: (ParameterTest_combination_all_Test::TestBody()+0x14f4) [0x560960]
124:  6: (void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x74) [0x5bbeac]
124:  7: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x68) [0x5b0f44]
124:  8: (testing::Test::Run()+0x110) [0x5848b0]
124:  9: (testing::TestInfo::Run()+0x140) [0x585368]
124:  10: (testing::TestSuite::Run()+0x14c) [0x585c60]
124:  11: (testing::internal::UnitTestImpl::RunAllTests()+0x480) [0x5929c0]
124:  12: (bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x74) [0x5bd978]
124:  13: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x68) [0x5b2734]
124:  14: (testing::UnitTest::Run()+0xe4) [0x590f84]
124:  15: (RUN_ALL_TESTS()+0x18) [0x5652ec]
124:  16: (main()+0xb8) [0x563248]
124:  17: (__libc_start_main()+0x114) [0xb01d0848]
124: 
124:  ceph version 15.2.4 (7447c15c6ff58d7fce91843b705a268a1917325c) octopus (stable)
124:  1: (()+0x14f564) [0x5f1564]
124:  2: (__default_sa_restorer()+0) [0xb01e6400]
124: 2020-11-25T07:14:04.634+0000 b6f1f010 -1 *** Caught signal (Aborted) **
124:  in thread b6f1f010 thread_name:unittest_erasur
124: 
124:  ceph version 15.2.4 (7447c15c6ff58d7fce91843b705a268a1917325c) octopus (stable)
124:  1: (()+0x14f564) [0x5f1564]
124:  2: (__default_sa_restorer()+0) [0xb01e6400]
124:  NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
124: 
124: --- begin dump of recent events ---
124:    -29> 2020-11-25T07:14:04.504+0000 b6f1f010  5 asok(0x683840) register_command assert hook 0x66be20
124:    -28> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command abort hook 0x66be20
124:    -27> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command leak_some_memory hook 0x66be20
124:    -26> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perfcounters_dump hook 0x66be20
124:    -25> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command 1 hook 0x66be20
124:    -24> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perf dump hook 0x66be20
124:    -23> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perfcounters_schema hook 0x66be20
124:    -22> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perf histogram dump hook 0x66be20
124:    -21> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command 2 hook 0x66be20
124:    -20> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perf schema hook 0x66be20
124:    -19> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perf histogram schema hook 0x66be20
124:    -18> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perf reset hook 0x66be20
124:    -17> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config show hook 0x66be20
124:    -16> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config help hook 0x66be20
124:    -15> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config set hook 0x66be20
124:    -14> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config unset hook 0x66be20
124:    -13> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config get hook 0x66be20
124:    -12> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config diff hook 0x66be20
124:    -11> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config diff get hook 0x66be20
124:    -10> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command injectargs hook 0x66be20
124:     -9> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command log flush hook 0x66be20
124:     -8> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command log dump hook 0x66be20
124:     -7> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command log reopen hook 0x66be20
124:     -6> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command dump_mempools hook 0x6a61a4
124:     -5> 2020-11-25T07:14:04.539+0000 b6f1f010 -1 Errors while parsing config file!
124:     -4> 2020-11-25T07:14:04.539+0000 b6f1f010 -1 parse_file: filesystem error: cannot get file size [ceph.conf]
124:     -3> 2020-11-25T07:14:04.539+0000 b6f1f010 -1 Errors while parsing config file!
124:     -2> 2020-11-25T07:14:04.539+0000 b6f1f010 -1 parse_file: filesystem error: cannot get file size [ceph.conf]
124:     -1> 2020-11-25T07:14:04.634+0000 b6f1f010 -1 /mnt/git/debug/ceph-15.2.4/src/erasure-code/shec/ErasureCodeShec.cc: In function 'virtual int ErasureCodeShec::_decode(const std::set<int>&, const std::map<int, ceph::buffer::v15_2_0::list>&, std::map<int, ceph::buffer::v15_2_0::list>*)' thread b6f1f010 time 2020-11-25T07:14:04.613041+0000
124: /mnt/git/debug/ceph-15.2.4/src/erasure-code/shec/ErasureCodeShec.cc: 200: FAILED ceph_assert(!chunks.empty())
124: 
124:  ceph version 15.2.4 (7447c15c6ff58d7fce91843b705a268a1917325c) octopus (stable)
124:  1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x190) [0xb1eb8df4]
124:  2: (()+0x1551090) [0xb1eb9090]
124:  3: (ErasureCodeShec::_decode(std::set<int, std::less<int>, std::allocator<int> > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > >*)+0x2a4) [0xb09277f0]
124:  4: (ceph::ErasureCode::decode(std::set<int, std::less<int>, std::allocator<int> > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > >*, int)+0x40) [0xb090ec44]
124:  5: (ParameterTest_combination_all_Test::TestBody()+0x14f4) [0x560960]
124:  6: (void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x74) [0x5bbeac]
124:  7: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x68) [0x5b0f44]
124:  8: (testing::Test::Run()+0x110) [0x5848b0]
124:  9: (testing::TestInfo::Run()+0x140) [0x585368]
124:  10: (testing::TestSuite::Run()+0x14c) [0x585c60]
124:  11: (testing::internal::UnitTestImpl::RunAllTests()+0x480) [0x5929c0]
124:  12: (bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x74) [0x5bd978]
124:  13: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x68) [0x5b2734]
124:  14: (testing::UnitTest::Run()+0xe4) [0x590f84]
124:  15: (RUN_ALL_TESTS()+0x18) [0x5652ec]
124:  16: (main()+0xb8) [0x563248]
124:  17: (__libc_start_main()+0x114) [0xb01d0848]
124: 
124:      0> 2020-11-25T07:14:04.634+0000 b6f1f010 -1 *** Caught signal (Aborted) **
124:  in thread b6f1f010 thread_name:unittest_erasur
124: 
124:  ceph version 15.2.4 (7447c15c6ff58d7fce91843b705a268a1917325c) octopus (stable)
124:  1: (()+0x14f564) [0x5f1564]
124:  2: (__default_sa_restorer()+0) [0xb01e6400]
124:  NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
124: 
124: --- logging levels ---
124:    0/ 5 none
124:    0/ 1 lockdep
124:    0/ 1 context
124:    1/ 1 crush
124:    1/ 5 mds
124:    1/ 5 mds_balancer
124:    1/ 5 mds_locker
124:    1/ 5 mds_log
124:    1/ 5 mds_log_expire
124:    1/ 5 mds_migrator
124:    0/ 1 buffer
124:    0/ 1 timer
124:    0/ 1 filer
124:    0/ 1 striper
124:    0/ 1 objecter
124:    0/ 5 rados
124:    0/ 5 rbd
124:    0/ 5 rbd_mirror
124:    0/ 5 rbd_replay
124:    0/ 5 rbd_rwl
124:    0/ 5 journaler
124:    0/ 5 objectcacher
124:    0/ 5 immutable_obj_cache
124:    0/ 5 client
124:    1/ 5 osd
124:    0/ 5 optracker
124:    0/ 5 objclass
124:    1/ 3 filestore
124:    1/ 3 journal
124:    0/ 0 ms
124:    1/ 5 mon
124:    0/10 monc
124:    1/ 5 paxos
124:    0/ 5 tp
124:    1/ 5 auth
124:    1/ 5 crypto
124:    1/ 1 finisher
124:    1/ 1 reserver
124:    1/ 5 heartbeatmap
124:    1/ 5 perfcounter
124:    1/ 5 rgw
124:    1/ 5 rgw_sync
124:    1/10 civetweb
124:    1/ 5 javaclient
124:    1/ 5 asok
124:    1/ 1 throttle
124:    0/ 0 refs
124:    1/ 5 compressor
124:    1/ 5 bluestore
124:    1/ 5 bluefs
124:    1/ 3 bdev
124:    1/ 5 kstore
124:    4/ 5 rocksdb
124:    4/ 5 leveldb
124:    4/ 5 memdb
124:    1/ 5 fuse
124:    1/ 5 mgr
124:    1/ 5 mgrc
124:    1/ 5 dpdk
124:    1/ 5 eventtrace
124:    1/ 5 prioritycache
124:    0/ 5 test
124:   -2/-2 (syslog threshold)
124:   99/99 (stderr threshold)
124: --- pthread ID / name mapping for recent threads ---
124:   b6f1f010 / unittest_erasur
124:   max_recent       500
124:   max_new         1000
124:   log_file 
124: --- end dump of recent events ---
124: --- begin dump of recent events ---
124:    -29> 2020-11-25T07:14:04.504+0000 b6f1f010  5 asok(0x683840) register_command assert hook 0x66be20
124:    -28> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command abort hook 0x66be20
124:    -27> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command leak_some_memory hook 0x66be20
124:    -26> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perfcounters_dump hook 0x66be20
124:    -25> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command 1 hook 0x66be20
124:    -24> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perf dump hook 0x66be20
124:    -23> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perfcounters_schema hook 0x66be20
124:    -22> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perf histogram dump hook 0x66be20
124:    -21> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command 2 hook 0x66be20
124:    -20> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perf schema hook 0x66be20
124:    -19> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perf histogram schema hook 0x66be20
124:    -18> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command perf reset hook 0x66be20
124:    -17> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config show hook 0x66be20
124:    -16> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config help hook 0x66be20
124:    -15> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config set hook 0x66be20
124:    -14> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config unset hook 0x66be20
124:    -13> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config get hook 0x66be20
124:    -12> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config diff hook 0x66be20
124:    -11> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command config diff get hook 0x66be20
124:    -10> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command injectargs hook 0x66be20
124:     -9> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command log flush hook 0x66be20
124:     -8> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command log dump hook 0x66be20
124:     -7> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command log reopen hook 0x66be20
124:     -6> 2020-11-25T07:14:04.509+0000 b6f1f010  5 asok(0x683840) register_command dump_mempools hook 0x6a61a4
124:     -5> 2020-11-25T07:14:04.539+0000 b6f1f010 -1 Errors while parsing config file!
124:     -4> 2020-11-25T07:14:04.539+0000 b6f1f010 -1 parse_file: filesystem error: cannot get file size [ceph.conf]
124:     -3> 2020-11-25T07:14:04.539+0000 b6f1f010 -1 Errors while parsing config file!
124:     -2> 2020-11-25T07:14:04.539+0000 b6f1f010 -1 parse_file: filesystem error: cannot get file size [ceph.conf]
124:     -1> 2020-11-25T07:14:04.634+0000 b6f1f010 -1 /mnt/git/debug/ceph-15.2.4/src/erasure-code/shec/ErasureCodeShec.cc: In function 'virtual int ErasureCodeShec::_decode(const std::set<int>&, const std::map<int, ceph::buffer::v15_2_0::list>&, std::map<int, ceph::buffer::v15_2_0::list>*)' thread b6f1f010 time 2020-11-25T07:14:04.613041+0000
124: /mnt/git/debug/ceph-15.2.4/src/erasure-code/shec/ErasureCodeShec.cc: 200: FAILED ceph_assert(!chunks.empty())
124: 
124:  ceph version 15.2.4 (7447c15c6ff58d7fce91843b705a268a1917325c) octopus (stable)
124:  1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x190) [0xb1eb8df4]
124:  2: (()+0x1551090) [0xb1eb9090]
124:  3: (ErasureCodeShec::_decode(std::set<int, std::less<int>, std::allocator<int> > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > >*)+0x2a4) [0xb09277f0]
124:  4: (ceph::ErasureCode::decode(std::set<int, std::less<int>, std::allocator<int> > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > > const&, std::map<int, ceph::buffer::v15_2_0::list, std::less<int>, std::allocator<std::pair<int const, ceph::buffer::v15_2_0::list> > >*, int)+0x40) [0xb090ec44]
124:  5: (ParameterTest_combination_all_Test::TestBody()+0x14f4) [0x560960]
124:  6: (void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x74) [0x5bbeac]
124:  7: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x68) [0x5b0f44]
124:  8: (testing::Test::Run()+0x110) [0x5848b0]
124:  9: (testing::TestInfo::Run()+0x140) [0x585368]
124:  10: (testing::TestSuite::Run()+0x14c) [0x585c60]
124:  11: (testing::internal::UnitTestImpl::RunAllTests()+0x480) [0x5929c0]
124:  12: (bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x74) [0x5bd978]
124:  13: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x68) [0x5b2734]
124:  14: (testing::UnitTest::Run()+0xe4) [0x590f84]
124:  15: (RUN_ALL_TESTS()+0x18) [0x5652ec]
124:  16: (main()+0xb8) [0x563248]
124:  17: (__libc_start_main()+0x114) [0xb01d0848]
124: 
124:      0> 2020-11-25T07:14:04.634+0000 b6f1f010 -1 *** Caught signal (Aborted) **
124:  in thread b6f1f010 thread_name:unittest_erasur
124: 
124:  ceph version 15.2.4 (7447c15c6ff58d7fce91843b705a268a1917325c) octopus (stable)
124:  1: (()+0x14f564) [0x5f1564]
124:  2: (__default_sa_restorer()+0) [0xb01e6400]
124:  NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
124: 
124: --- logging levels ---
124:    0/ 5 none
124:    0/ 1 lockdep
124:    0/ 1 context
124:    1/ 1 crush
124:    1/ 5 mds
124:    1/ 5 mds_balancer
124:    1/ 5 mds_locker
124:    1/ 5 mds_log
124:    1/ 5 mds_log_expire
124:    1/ 5 mds_migrator
124:    0/ 1 buffer
124:    0/ 1 timer
124:    0/ 1 filer
124:    0/ 1 striper
124:    0/ 1 objecter
124:    0/ 5 rados
124:    0/ 5 rbd
124:    0/ 5 rbd_mirror
124:    0/ 5 rbd_replay
124:    0/ 5 rbd_rwl
124:    0/ 5 journaler
124:    0/ 5 objectcacher
124:    0/ 5 immutable_obj_cache
124:    0/ 5 client
124:    1/ 5 osd
124:    0/ 5 optracker
124:    0/ 5 objclass
124:    1/ 3 filestore
124:    1/ 3 journal
124:    0/ 0 ms
124:    1/ 5 mon
124:    0/10 monc
124:    1/ 5 paxos
124:    0/ 5 tp
124:    1/ 5 auth
124:    1/ 5 crypto
124:    1/ 1 finisher
124:    1/ 1 reserver
124:    1/ 5 heartbeatmap
124:    1/ 5 perfcounter
124:    1/ 5 rgw
124:    1/ 5 rgw_sync
124:    1/10 civetweb
124:    1/ 5 javaclient
124:    1/ 5 asok
124:    1/ 1 throttle
124:    0/ 0 refs
124:    1/ 5 compressor
124:    1/ 5 bluestore
124:    1/ 5 bluefs
124:    1/ 3 bdev
124:    1/ 5 kstore
124:    4/ 5 rocksdb
124:    4/ 5 leveldb
124:    4/ 5 memdb
124:    1/ 5 fuse
124:    1/ 5 mgr
124:    1/ 5 mgrc
124:    1/ 5 dpdk
124:    1/ 5 eventtrace
124:    1/ 5 prioritycache
124:    0/ 5 test
124:   -2/-2 (syslog threshold)
124:   99/99 (stderr threshold)
124: --- pthread ID / name mapping for recent threads ---
124:   b6f1f010 / unittest_erasur
124:   max_recent       500
124:   max_new         1000
124:   log_file /var/lib/ceph/crash/2020-11-25T07:14:04.639325Z_023382f1-8190-4d3c-b87d-f02397f8a063/log
124: --- end dump of recent events ---
1/1 Test #124: unittest_erasure_code_shec_arguments ...Child aborted***Exception:   9.63 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   9.99 sec

The following tests FAILED:
	124 - unittest_erasure_code_shec_arguments (Child aborted)
Errors while running CTest

```

Reply:
Debug backtrace. ceph::ErasureCode::decode() was called with empty chunks from the test. May be the test itself is buggy but certainly ceph::ErasureCode::decode() should not fail when empty chunks is provided.

```
Thread 1 "unittest_erasur" hit Breakpoint 2, ErasureCodeShec::_decode (this=0x5cfc60, want_to_read=std::set with 1 element = {...}, chunks=std::map with 0 elements, 
    decoded=0xbefff7e0) at /mnt/git/debug/ceph-15.2.4/src/erasure-code/shec/ErasureCodeShec.cc:200
200	  ceph_assert(!chunks.empty());
(gdb) backtrace
#0  ErasureCodeShec::_decode (this=0x5cfc60, want_to_read=std::set with 1 element = {...}, chunks=std::map with 0 elements, decoded=0xbefff7e0)
    at /mnt/git/debug/ceph-15.2.4/src/erasure-code/shec/ErasureCodeShec.cc:200
#1  0xb09ddc44 in ceph::ErasureCode::decode (this=0x5cfc60, want_to_read=std::set with 1 element = {...}, chunks=std::map with 0 elements, decoded=0xbefff7e0, 
    chunk_size=32) at /mnt/git/debug/ceph-15.2.4/src/erasure-code/ErasureCode.cc:254
#2  0x004be960 in ParameterTest_combination_all_Test::TestBody (this=0x5a5818) at /mnt/git/debug/ceph-15.2.4/src/test/erasure-code/TestErasureCodeShec_arguments.cc:277
#3  0x00519eac in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (object=0x5a5818, method=&virtual testing::Test::TestBody(), 
    location=0x55aa4c "the test body") at /mnt/git/debug/ceph-15.2.4/src/googletest/googletest/src/gtest.cc:2593
#4  0x0050ef44 in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=0x5a5818, method=&virtual testing::Test::TestBody(), 
    location=0x55aa4c "the test body") at /mnt/git/debug/ceph-15.2.4/src/googletest/googletest/src/gtest.cc:2629
#5  0x004e28b0 in testing::Test::Run (this=0x5a5818) at /mnt/git/debug/ceph-15.2.4/src/googletest/googletest/src/gtest.cc:2668
#6  0x004e3368 in testing::TestInfo::Run (this=0x5dd138) at /mnt/git/debug/ceph-15.2.4/src/googletest/googletest/src/gtest.cc:2845
#7  0x004e3c60 in testing::TestSuite::Run (this=0x5d6160) at /mnt/git/debug/ceph-15.2.4/src/googletest/googletest/src/gtest.cc:2977
#8  0x004f09c0 in testing::internal::UnitTestImpl::RunAllTests (this=0x5d1d58) at /mnt/git/debug/ceph-15.2.4/src/googletest/googletest/src/gtest.cc:5518
#9  0x0051b978 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x5d1d58, 
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x4f0540 <testing::internal::UnitTestImpl::RunAllTests()>, 
    location=0x55b46c "auxiliary test code (environments or event listeners)") at /mnt/git/debug/ceph-15.2.4/src/googletest/googletest/src/gtest.cc:2593
#10 0x00510734 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x5d1d58, 
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x4f0540 <testing::internal::UnitTestImpl::RunAllTests()>, 
    location=0x55b46c "auxiliary test code (environments or event listeners)") at /mnt/git/debug/ceph-15.2.4/src/googletest/googletest/src/gtest.cc:2629
#11 0x004eef84 in testing::UnitTest::Run (this=0x589c20 <testing::UnitTest::GetInstance()::instance>)
    at /mnt/git/debug/ceph-15.2.4/src/googletest/googletest/src/gtest.cc:5105
#12 0x004c32ec in RUN_ALL_TESTS () at /mnt/git/debug/ceph-15.2.4/src/googletest/googletest/include/gtest/gtest.h:2472
#13 0x004c1248 in main (argc=1, argv=0xbefffc84) at /mnt/git/debug/ceph-15.2.4/src/test/erasure-code/TestErasureCodeShec_arguments.cc:407

```

Reply:
> Debug backtrace. ceph::ErasureCode::decode() was called with empty chunks from the test. May be the test itself is buggy but certainly ceph::ErasureCode::decode() should not fail when empty chunks is provided.

it depends on the our expectation to this function. i think we *assume* that the `chunks` feed to `decode()` should fulfill the needs to decode `want_to_read`.

could you paste the output of `info locals` in gdb in `TestErasureCodeShec_arguments.cc` when assert() fails?

Reply:
```
(gdb) info locals
assert_data_ctx = {assertion = 0xb0a173cc "!chunks.empty()", file = 0xb0a1733c "/mnt/git/debug/ceph-15.2.4/src/erasure-code/shec/ErasureCodeShec.cc", line = 200, 
  function = 0xb0a173dc "virtual int ErasureCodeShec::_decode(const std::set<int>&, const std::map<int, ceph::buffer::v15_2_0::list>&, std::map<int, ceph::buffer::v15_2_0::list>*)"}
have = std::vector of length 0, capacity 0
k = 4
m = 3
__PRETTY_FUNCTION__ = "virtual int ErasureCodeShec::_decode(const std::set<int>&, const std::map<int, ceph::buffer::v15_2_0::list>&, std::map<int, ceph::buffer::v15_2_0::list>*)"
blocksize = 6093920

```

Reply:
thank you! will look into this tomorrow.

Reply:
@vbashkirtsev the frame in which you printed `locals` was not in `TestErasureCodeShec_arguments.cc`. it was in `ErasureCodeShec.cc`. could you please do this in `TestErasureCodeShec_arguments.cc`? i'd like to understand the values of, for instance `want_to_read` and `array_available_chunks`.

Reply:
Ouch... I did a really stupid thing - locals above of course a plainly useless... :(

On iteration 129 in TestErasureCodeShec_arguments.cc:277 (just before failure):

```
Thread 1 "unittest_erasur" hit Breakpoint 2, ParameterTest_combination_all_Test::TestBody (this=0x5a5818)
    at /mnt/git/debug/ceph-15.2.4/src/test/erasure-code/TestErasureCodeShec_arguments.cc:277
277	          dresult = shec->decode(want_to_read, inchunks, &decoded,
(gdb) info locals
r2 = 0
w2 = 0
r1 = 0
w1 = 0
want_to_read_without_avails = std::set with 0 elements
minimum_chunks = std::map with 0 elements
itr = non-dereferenceable iterator for associative container
cmp = 1
minimum_count = 0
available_chunks = std::set with 0 elements
array_available_chunks = {0, 0, 0, 0, 0, 0, 0}
inchunks = std::map with 0 elements
decoded = std::map with 0 elements
array_want_to_read = {0, 0, 0, 0, 0, 0, 0}
dresult = -1340058416
usable = {_buffers = {_root = {next = 0xbefff724}, _tail = 0xbefff724}, _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 0, _num = 0, 
  static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}
out1 = {_buffers = {_root = {next = 0x63e610}, _tail = 0x63ef40}, _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 224, _num = 7, 
  static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}
want_to_encode = std::set with 7 elements = {[0] = 0, [1] = 1, [2] = 2, [3] = 3, [4] = 4, [5] = 5, [6] = 6}
encoded = std::map with 7 elements = {[0] = {_buffers = {_root = {next = 0x63e658}, _tail = 0x63e658}, 
    _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 32, _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [1] = {
    _buffers = {_root = {next = 0x63e708}, _tail = 0x63e708}, _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 32, _num = 1, 
    static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [2] = {_buffers = {_root = {next = 0x63e7e8}, _tail = 0x63e7e8}, 
    _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 32, _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [3] = {
    _buffers = {_root = {next = 0x63e8c8}, _tail = 0x63e8c8}, _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 32, _num = 1, 
    static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [4] = {_buffers = {_root = {next = 0x63e9a8}, _tail = 0x63e9a8}, _carriage = 0x63e9ac, _len = 32, 
    _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [5] = {_buffers = {_root = {next = 0x63eb50}, _tail = 0x63eb50}, _carriage = 0x63eb54, 
    _len = 32, _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [6] = {_buffers = {_root = {next = 0x63ec30}, _tail = 0x63ec30}, 
    _carriage = 0x63ec34, _len = 32, _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}}
want_to_read = std::set with 0 elements
result = 0
alignment = 128
tail = 0
padded_length = 128
kObjectSize = 128
k = 0x558b9c "4"
m = 0x558ba0 "3"
c = 0x558ba4 "2"
i_k = 4
i_m = 3
i_c = 2
c_size = 32
tcache = {_vptr.ErasureCodeShecTableCache = 0xb0a2fec8 <vtable for ErasureCodeShecTableCache+8>, static decoding_tables_lru_length = 10000, codec_tables_guard = {<ceph::mutex_debug_detail::mutex_debugging_base> = {group = "shec-lru-cache", id = -1, lockdep = true, backtrace = false, nlock = 0, locked_by = {_M_thread = 0}}, m = {__data = {__lock = 0, __count = 0, __owner = 0, __kind = 2, __nusers = 0, {__spins = 0, __list = {__next = 0x0}}}, __size = '\000' <repeats 12 times>, "\002\000\000\000\000\000\000\000\000\000\000", __align = 0}, static recursive = false}, encoding_table = std::map with 1 element = {[0] = std::map with 1 element = {[4] = std::map with 1 element = {[3] = std::map with 1 element = {[2] = std::map with 1 element = {[8] = 0x5d7ac8}}}}}, decoding_tables = std::map with 1 element = {[0] = 0x63ef78}, decoding_tables_lru = std::map with 1 element = {[0] = 0x63e830}}
shec = 0x5cfc60
profile = 0x608470
in = {_buffers = {_root = {next = 0x616d58}, _tail = 0x616d58}, _carriage = 0x616d5c, _len = 128, _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}
```

Reply:
@vbashkirtsev i am confused because after looking into the values of the local variable, i think the test is correct. i am quoting part of `ErasureCodeShec::_decode()` here:

```c++
  if (includes(
	have.begin(), have.end(), want_to_read.begin(), want_to_read.end())) {
    for (set<int>::iterator i = want_to_read.begin();
	 i != want_to_read.end();
	 ++i) {
      (*decoded)[*i] = chunks.find(*i)->second;
    }
    return 0;
  }
```

and quoting related local variables here

```
have = std::vector of length 0, capacity 0
want_to_read = std::set with 0 elements
```

and the implementation of `std::includes()` of libstdc++ here
```c++
  template<typename _InputIterator1, typename _InputIterator2,
       typename _Compare>
    bool
    __includes(_InputIterator1 __first1, _InputIterator1 __last1,
           _InputIterator2 __first2, _InputIterator2 __last2,
           _Compare __comp)
    {
      while (__first1 != __last1 && __first2 != __last2)
    if (__comp(__first2, __first1))
      return false;
    else if (__comp(__first1, __first2))
      ++__first1;
    else
      {
        ++__first1;
        ++__first2;
      }

      return __first2 == __last2;
    }
```

so `ErasureCodeShec::_decode()` should return "0" in this case, right?

Reply:
For completeness: I use GCC 9.1.0 and my /usr/include/c++/9.1.0/bits/stl_algo.h has the same code as you quoted above.

However nothing wrong with includes() guard above. In fact it is me who flunked in this very thread for second time and I should apologise profusely for misleading you: `info locals` I have posted above was one iteration BEFORE failure. Below are `info locals` just prior the failure (I have double checked this time that it is correct one).

```
r2 = 0
w2 = 0
r1 = 1
w1 = 1
want_to_read_without_avails = std::set with 0 elements
minimum_chunks = std::map with 0 elements
itr = non-dereferenceable iterator for associative container
cmp = 1
minimum_count = 0
available_chunks = std::set with 0 elements
array_available_chunks = {0, 0, 0, 0, 0, 0, 0}
inchunks = std::map with 0 elements
decoded = std::map with 0 elements
array_want_to_read = {1, 0, 0, 0, 0, 0, 0}
dresult = 0
usable = {_buffers = {_root = {next = 0xbefff724}, _tail = 0xbefff724}, _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 0, _num = 0, 
  static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}
out1 = {_buffers = {_root = {next = 0x63e610}, _tail = 0x63ef40}, _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 224, _num = 7, 
  static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}
want_to_encode = std::set with 7 elements = {[0] = 0, [1] = 1, [2] = 2, [3] = 3, [4] = 4, [5] = 5, [6] = 6}
encoded = std::map with 7 elements = {[0] = {_buffers = {_root = {next = 0x63e658}, _tail = 0x63e658}, 
    _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 32, _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [1] = {
    _buffers = {_root = {next = 0x63e708}, _tail = 0x63e708}, _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 32, _num = 1, 
    static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [2] = {_buffers = {_root = {next = 0x63e7e8}, _tail = 0x63e7e8}, 
    _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 32, _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [3] = {
    _buffers = {_root = {next = 0x63e8c8}, _tail = 0x63e8c8}, _carriage = 0xb6f3261c <ceph::buffer::v15_2_0::list::always_empty_bptr>, _len = 32, _num = 1, 
    static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [4] = {_buffers = {_root = {next = 0x63e9a8}, _tail = 0x63e9a8}, _carriage = 0x63e9ac, _len = 32, 
    _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [5] = {_buffers = {_root = {next = 0x63eb50}, _tail = 0x63eb50}, _carriage = 0x63eb54, 
    _len = 32, _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}, [6] = {_buffers = {_root = {next = 0x63ec30}, _tail = 0x63ec30}, 
    _carriage = 0x63ec34, _len = 32, _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}}
want_to_read = std::set with 1 element = {[0] = 0}
result = -5
alignment = 128
tail = 0
padded_length = 128
kObjectSize = 128
k = 0x558b9c "4"
m = 0x558ba0 "3"
c = 0x558ba4 "2"
i_k = 4
i_m = 3
i_c = 2
c_size = 32
tcache = {_vptr.ErasureCodeShecTableCache = 0xb0a2fec8 <vtable for ErasureCodeShecTableCache+8>, static decoding_tables_lru_length = 10000, codec_tables_guard = {<ceph::mutex_debug_detail::mutex_debugging_base> = {group = "shec-lru-cache", id = -1, lockdep = true, backtrace = false, nlock = 0, locked_by = {_M_thread = 0}}, m = {__data = {__lock = 0, __count = 0, __owner = 0, __kind = 2, __nusers = 0, {__spins = 0, __list = {__next = 0x0}}}, __size = '\000' <repeats 12 times>, "\002\000\000\000\000\000\000\000\000\000\000", __align = 0}, static recursive = false}, encoding_table = std::map with 1 element = {[0] = std::map with 1 element = {[4] = std::map with 1 element = {[3] = std::map with 1 element = {[2] = std::map with 1 element = {[8] = 0x5d7ac8}}}}}, decoding_tables = std::map with 1 element = {[0] = 0x63ef78}, decoding_tables_lru = std::map with 1 element = {[0] = 0x63e830}}
shec = 0x5cfc60
profile = 0x608470
in = {_buffers = {_root = {next = 0x616d58}, _tail = 0x616d58}, _carriage = 0x616d5c, _len = 128, _num = 1, static always_empty_bptr = {_raw = 0x0, _off = 0, _len = 0}}

```

Following is passed to decode():

`want_to_read = std::set with 1 element = {[0] = 0}`

while

`available_chunks = std::set with 0 elements`

which makes me think that the test is incorrect. But even though the test may be faulty _decode() should not spit the dummy just because someone passed bad parameters to it and guard checking chunks.size() is warranted and appears to be a proper fix.

Reply:
@tchaikov So what do you think about this patch? Should it be adopted?

Reply:
@vbashkirtsev i think we should handle this as a special case, and bail out early. also, i collect the bits i proposed in this PR in #38665.